### ISSUE TITLE
[V3] timeout validated as port number

### ISSUE THREAD

**Issue by individual-it** [pact-foundation/pact-js#761]
### Software versions

- **OS**: Ubuntu 21.04
- **Consumer Pact library**: Pact JS 10.0.0-beta.51
- **Provider Pact library**: Pact JS 10.0.0-beta.51
- **Node Version**: `v14.18.0`

### Issue Checklist

Please confirm the following:

- [x] I have upgraded to the latest
- [x] I have the read the FAQs in the Readme
- [x] I have triple checked, that there are **no unhandled promises** in my code and have [read](https://github.com/pact-foundation/pact-js/#test-fails-when-it-should-pass) the section on intermittent test failures
- [x] I have set my log level to debug and attached a log file showing the complete request/response cycle
- [x] For bonus points and virtual high fives, I have created a reproduceable git repository (see below) to illustrate the problem

### Expected behaviour

timeout should be validated as int not as port number

### Actual behaviour

timeout cannot be set to anything higher than `65535` and the error message says its not a valid port-number

### Steps to reproduce
1. set `timeout` in the provider test configuration to `65536`
2. run the tests

https://github.com/individual-it/pact-js/commit/d23add2041e4b90f979fd135fdd9de9ed396e83d

### Relevant log files

```
[2021-10-20T12:21:49Z ERROR pact_ffi::verifier::verifier] error verifying Pact: "error: Invalid value for '--request-timeout <request-timeout>': '65536' is not a valid port value: number too large to fit in target type" Error { message: "error: Invalid value for '--request-timeout <request-timeout>': '65536' is not a valid port value: number too large to fit in target type", kind: ValueValidation, info: None }
[2021-10-20 12:21:49.776 +0000] ERROR (39243 on deepthought3): pact-core@13.2.1: !!!!!!!!! PACT CRASHED !!!!!!!!!
```

**Comment by diestrin** [comment]
Any updates on this bug? I have been dealing with some timeouts, increasing it might work, but this is preventing it to happen

**Comment by diestrin** [comment]
From what I can see it is still the same https://github.com/pact-foundation/pact-reference/blob/bbf8947dff2d98d59e5e29e3000d152d77be6a5f/rust/pact_verifier_cli/src/args.rs#L5
@mefellows did this got updated somewhere else?

**Comment by mefellows** [comment]
Hi @diestrin , the situation is a little more complicated, we've been having a few issues with `ref-napi` causing intermittent build failures, owing to a structural problem in that library. This has stalled the momentum in the V3 upgrade.

We've since ported `pact-js-core` to a node gyp project with direct bindings to the underlying c library (the rust project above). 

TL;DR - we need to fix Rust, but it will be a little bit before it's made available to this project sadly.

**Comment by mefellows** [comment]
See also https://github.com/pact-foundation/pact-reference/pull/175.

**Comment by diestrin** [comment]
I understand. Thanks for fixing. Do you have a guess of how much time are we talking? It stuck my pact tests, I need provider states, so I really needed V3 upgrade, but the tests are increasing and I need much more than 65 seconds to complete the whole test run.

I saw before there was a [timeout per state handler](https://github.com/pact-foundation/pact-js/blob/e891fccb5802491cf148f398c81406bf06ae43c8/src/v3/verifier.ts#L22) in the V3 provider, but now the options is gone? It seems weird to me that I have to specify now just a single timeout for the whole run. With too many cases I'll have to bump the timeout number to the limits, and right now the maximum I can set it is 65 seconds.

Thanks for the help!

**Comment by mefellows** [comment]
OK, there are a few problems.

1. The timeout value passed in as the options is used in two places - in the rust core, it's a _per-request_ timeout (i.e. each state handler invocation and execution of the actual test)
2. The timeout value is use by the `verify` [method](https://github.com/pact-foundation/pact-js-core/blob/master/src/verifier/index.ts#L24) in the core, which wraps the entire test process.
3. Because of 1/2, the 16 bit integer problem is manifesting itself more often, because it's used for both purposes. It's unlikely a per-request timeout would actually need to exceed a 16 bit integer, but we probably should still support it.

This can be addressed in a number of ways:

1. We add a new option that separates the overall timeout from the per-request one (I'd propose we keep `timeout` as the total timeout and a new option for the per-request one)
2. We remove the code from the core that does an overall timeout, considering that functionality is trivially managed outside of the core

I'm in favour of (2) at this time, but keen on other opinions here.

>  It seems weird to me that I have to specify now just a single timeout for the whole run

I think you're possibly arguing the same case.

> Do you have a guess of how much time are we talking?

I think I could get another v3 beta to you in the next week or so, albeit wrapping up the v3 work is likely to take a few months at this present moment. There's possibly nothing stopping us releasing a new beta version, given it's...a beta.

**Comment by TimothyJones** [comment]
> It's unlikely a per-request timeout would actually need to exceed a 16 bit integer, but we probably should still support it.

What's the reason to use a 16 bit integer? I doubt it would lead to any space saving on modern architecture.

>> We remove the code from the core that does an overall timeout, considering that functionality is trivially managed outside of the core
> I'm in favour of (2) at this time, but keen on other opinions here.

I agree. I think the use case for an overall timeout isn't clear (you'd want to always set it to "very long", or something like "N times the number of test cases"). Also, if there is a `timeout` option without any other information, I would naturally expect it to be per test.

**Comment by mefellows** [comment]
> What's the reason to use a 16 bit integer? I doubt it would lead to any space saving on modern architecture.

I think it's the classic linux port range max of 65535. 

> I agree. I think the use case for an overall timeout isn't clear (you'd want to always set it to "very long", or something like "N times the number of test cases"). Also, if there is a timeout option without any other information, I would naturally expect it to be per test.

Cool, let's make it so!

**Comment by mefellows** [comment]
Reopening, because merging the rust code doesn't actually fix this :P

**Comment by diestrin** [comment]
I guess an overall timeout is to avoid having a test stuck and running for 2 hours, consuming CI/CD minutes. But in my opinion, it is responsibility of the operator to make sure this doesn't happen. I don't have an overall timeout for other test tools I'm using neither, they all assume consumers will deal will that.

So, I'm all in for per-requests timeouts. Thanks again.

**Comment by mefellows** [comment]
Hi @diestrin, I've pushed out a new beta (v10.0.0-beta.56) with the timeout removed - would you mind please seeing if this issue has now been resolved?

**Comment by diestrin** [comment]
@mefellows I tried the version, but it gives me and error in the given method:
```
PactffiGivenWithParam(arg 2) expected a string

      68 |       .given(`existing users`, { users, current, synonyms })
    > 69 |       .uponReceiving(`a request for matches for ${ctx.$id}`)
         |        ^
```

Do you know if there was another change in those parameters?

**Comment by mefellows** [comment]
Awesome - closing this off 🎉
